### ISSUE TITLE
Debian package structure and pull-upstream-changes process

### ISSUE THREAD

**Issue by danpovey** [son-of-gridengine/sge#5]
https://salsa.debian.org/hpc-team/gridengine
@0xaf1f, I'd like to understand how that repo relates to this repo and how you pulled in changes.
In that repo, you seem to have the stuff from this repo, but also a debian/ subdirectory that contains a copy of sources/.  Is that just a copy of the top-level sources/?

**Comment by 0xaf1f** [comment]
> I'd like to understand how that repo relates to this repo and how you pulled in changes.

It's mostly an import of the release tarball. Only the `debian` directory is what I directly work on.

> In that repo, you seem to have the stuff from this repo, but also a debian/ subdirectory that contains a copy of sources/. Is that just a copy of the top-level sources/?

`debian/source` is unrelated to the top-level `source` directory. `debian/source` contains usually just one file to define the Debian package format.

> do you use the -spool-classic option when installing for Debian, or it still supports berkeleydb? I seem to remember there were hard-looking compilation errors to solve without the '-spool-classic' option.

It uses berkeleydb. This is the part of the packaging that runs aimk:

https://salsa.debian.org/hpc-team/gridengine/blob/a95a253e20bcbb32dc695b464d75ad758b842ab1/debian/rules#L39

Although the Debian package currently doesn't build in Unstable right now because of https://bugs.debian.org/cgi-bin/bugreport.cgi?bug=875404

**Comment by danpovey** [comment]
Thanks a lot for the info!  I notice the debian/patches/ subdirectory contains a number of small patches, mostly to the build system.  Are those already applied in your repo? 

I was investigating the ways in which your Debian repo differs from this repo.
Your Debian repo (based on the commit messages) seems to think it's up to date for upstream 8.1.9, and your 'master' branch is very close to the tag '8.1.9+dfsg'.  However, when I compare the code with our current upstream code for 8.1.9, I see differences.

For instance, in your
source/daemons/shepherd/builtin_starter.c
there is some old code referencing 'pre_args_ptr' that was removed in the upstream.
(it was removed in commit 9a9fe65fd15ce948295dcd62cc29b23ca471fa73 in this repo).

So I'm wondering whether your 8.1.9 is maybe not fully up-to-date -- perhaps you imported something that was later changed?

**Comment by 0xaf1f** [comment]
> I notice the debian/patches/ subdirectory contains a number of small patches, mostly to the build system. Are those already applied in your repo?

No-- everything outside of the `debian/` directory is unmodified from the release. During the package build process, the patches in `debian/patches` are applied and then the build is started.

> I was investigating the ways in which your Debian repo differs from this repo.
Your Debian repo (based on the commit messages) seems to think it's up to date for upstream 8.1.9, and your 'master' branch is very close to the tag '8.1.9+dfsg'. However, when I compare the code with our current upstream code for 8.1.9, I see differences.

That's odd. I took directly from the release tarball and only removed a few PDFs without source (hence the `+dfsg` suffix), so maybe the tag was off. The orig.tar.gz tarball from Debian (http://http.debian.net/debian/pool/main/g/gridengine/gridengine_8.1.9+dfsg.orig.tar.gz) is the pristine upstream source minus the files excluded (and those files are named at the top `debian/copyright` in the source package).

The original place they were downloadable from is gone now, but it seems that Dave had put them up on sourceforge later on: https://sourceforge.net/projects/gridengine/files/SGE/releases/ . This release tarball for 8.1.9: https://sourceforge.net/projects/gridengine/files/SGE/releases/8.1.9/sge-8.1.9.tar.gz/download is probably it.

**Comment by 0xaf1f** [comment]
I don't think it's useful to examine the git history of the upstream source in the Debian repository since it's just a series of imports from a tarball, at least since I started maintaining it. So regardless of what happened between 8.1.3 and 8.1.4, whatever exists in the 8.1.9 release tarball should be in our packages because everything got replaced by that.

However, I recently came across another repository of Dave's, which is a linear release repository: https://gitlab.com/loveshack/sge-release. I think this is what @bodgerer was talking about. At that tag for 8.1.9, line 212 of `source/daemons/execd/msg_execd.h` has

```
#define MSG_SYSTEM_CANTOPENTMPDIR_S   _MESSAGE(29124, _("can't open tmpdir "SFN))
```

**Comment by 0xaf1f** [comment]
by the way, there are repos for the SGE components too: https://gitlab.com/users/loveshack/projects

**Comment by 0xaf1f** [comment]
>  At that tag for 8.1.9, line 212 of source/daemons/execd/msg_execd.h has
>
> `#define MSG_SYSTEM_CANTOPENTMPDIR_S   _MESSAGE(29124, _("can't open tmpdir "SFN))`

Here's the exact link in gitlab: https://gitlab.com/loveshack/sge-release/blob/819/source/daemons/execd/msg_execd.h#L212

**Comment by danpovey** [comment]
OK, when I get a chance I'll have to look at how his 2 repos differ and we
can decide which one it makes more sense to start from.

**Comment by Kunzol** [comment]
As far as I remember only Version 8.1.9 can be used for Debian Stretch because of OpenSSL version.

See [Ticket 1572 ](https://arc.liv.ac.uk/trac/SGE/ticket/1572) 

and [Ticket 1618](https://arc.liv.ac.uk/trac/SGE/ticket/1618)

I am not sure, if it is a good idea to change the current code to make it build the packages for official Debian repos. It has some advantages to have all installed in one directory tree, e.g. distribution via network filesystem. At least there should be a configuration setting to have a choice.

**Comment by danpovey** [comment]
Marco: I wasn't proposing to change the current code to build the debian
packages-- well, certainly not by default.  I just wanted some
clarification on the Debian build process, mostly because I wanted to
figure out how we can test changes.
